format/spec.md

```diff
@@ -486,16 +486,17 @@ When reading v1 manifests with no sequence number column, sequence numbers for a
-| _optional_ | _required_ | **`manifest-list`**      | The location of a manifest list for this snapshot that tracks manifest files with additional metadata |
-| _optional_ |            | **`manifests`**          | A list of manifest file locations. Must be omitted if `manifest-list` is present |
-| _optional_ | _required_ | **`summary`**            | A string map that summarizes the snapshot changes, including `operation` (see below) |
-| _optional_ | _optional_ | **`schema-id`**          | ID of the table's current schema when the snapshot was created |
+| v1         | v2         | Field                    | Description                                                                                                                                                     |
+| ---------- | ---------- | ------------------------ |-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| _required_ | _required_ | **`snapshot-id`**        | A unique long ID                                                                                                                                                |
+| _optional_ | _optional_ | **`parent-snapshot-id`** | The snapshot ID of the snapshot's parent. Omitted for any snapshot with no parent                                                                               |
+|            | _required_ | **`sequence-number`**    | A monotonically increasing long that tracks the order of changes to a table                                                                                     |
+| _required_ | _required_ | **`timestamp-ms`**       | A timestamp when the snapshot was created, used for garbage collection and table inspection                                                                     |
+| _optional_ | _required_ | **`manifest-list`**      | The location of a manifest list for this snapshot that tracks manifest files with additional metadata                                                           |
+| _optional_ |            | **`manifests`**          | A list of manifest file locations. Must be omitted if `manifest-list` is present                                                                                |
+| _optional_ | _required_ | **`summary`**            | A string map that summarizes the snapshot changes, including `operation` (see below)                                                                            |
+| _optional_ | _optional_ | **`schema-id`**          | ID of the table's current schema when the snapshot was created                                                                                                  |
+| _optional_ | _optional_ | **`statistics`**         | A [statistics file's metadata](#statistics-file). The field should be retained by writers, unless writer updates the statistics, or knows they became obsolete. |
```

Comment:
Do we need statistics in the snapshot level if we are going to "have one stats file for a table"?

Reply:
If there is only one statistics file per table, we can have it in [Table metadata](https://iceberg.apache.org/spec/#table-metadata) instead of [snapsthots](https://iceberg.apache.org/spec/#snapshots).
Is there a link why we want only one statistics file per table? I assume per-snapshot one makes more sense. For example, we can remove the statistics files of a snapshot when we expire the snapshot.

Reply:
@rdblue do you have any thoughts about this?

Reply:
This is a good point, but there are tradeoffs in both directions.

If we attach a stats file at the table level, then we don't need to add logic for carrying it through from one snapshot to the next. That's nice because older writers could be unaware of stats and not propagate the file. That means that clients would need to look through commit history to find the most recent stats file attached to an ancestor snapshot. In short, there's some compatibility work to do here if we use snapshot. On the other hand, if we attach stats to snapshots, then we can have stats at the branch level.

We could design a new table-level structure that tracks stats files per branch, something like this:

```json
"stats": [
    { "ref": "main", "last-updated-snapshot-id": ..., "stats-path": "...", "file-size-in-bytes": ..., "footer-size-in-bytes": ..., blob-metadata: [ ... ] },
    { "ref": "test", "last-updated-snapshot-id": ..., "stats-path": "...", "file-size-in-bytes": ..., "footer-size-in-bytes": ..., blob-metadata: [ ... ] }
  ]
```

I would probably prefer that new structure, or something similar. What do you think, @findepi and @flyrain?

Reply:
Trying to understand the table level stats files, assume the table only has one branch, so the stats file will be updated every time a new snapshot created. For example, instead of having (snapshot1, stats-file1), (snapshot2, stats-file2)... There is only one stats file. How does time travel work with one stats file?
With snapshot-level stats files, expireSnapshot and time travel would be more natural to handle. 
> That means that clients would need to look through commit history to find the most recent stats file attached to an ancestor snapshot. 

I understand the compatibility concern, and the snapshots created by the old client won't have the stats files. Maybe I'm missing something, but it shouldn't be so hard for the new client find the most recent stats files and carry them over, right?

Reply:
> it shouldn't be so hard for the new client find the most recent stats files and carry them over, right?

@flyrain 
with current implementation --
iceberg library will do this automatically.

however, clients using old Iceberg library will "forget" stats when creating new snapshot (eg inserting new data)

> I would probably prefer that new structure, or something similar. What do you think, @findepi and @flyrain?


@rdblue i am not fluent in Iceberg Table / Snapshot intricacies, lifecycles, and branching, so I don't have opinion.

I know what we need to be able to
- attach a stats file to a table 
- update/replace/remove existing stats file
- retain existing stats file when doing eg DML on a table 
- reading old snapshot (time travel) should be able to use the stat file as was present in the table at that time

i believe attaching the stats file to snapshot addresses these needs.
if there is some other and better way to address these needs, please be explicit. I still need guidance in Iceberg world.

Reply:
+1 for storing at snapshot level.

Reply:
After thinking about this more and talking with @danielcweeks, I think that stats files should be tracked by a new top-level `stats` field in `TableMetadata`.

The issues with tracking stats in a snapshot are:
1. There is now a rule about carrying forward stats files that all writers must implement. This makes it difficult to manage an environment with multiple writer versions. The stats files will be dropped by some writers, and there isn't a good way to detect whether omitting stats was intentional (they are no longer accurate) or accidental (the writer didn't know about stats).
2. Adding or updating stats creates a new snapshot. The new snapshot has no data changes, but will cause retries and validations for concurrent writers.

A top-level `stats` field addresses the writer challenge because modifying stats files is separate from adding a snapshot and updating a ref. The catalog can detect the difference between removing a stats file (`stats` list exists) and accidentally dropping stats (`stats` list is omitted).

This also allows us to add to the Table API, rather than adding to the Snapshot API. The table API is better because matches are fuzzy -- there isn't necessarily a stats file that was calculated for a given snapshot so we don't want to associated a snapshot 1-to-1 with a stats file. (Side benefit: we don't copy all the stats metadata into each snapshot!)

The main drawback is that this would need to contain more than one stats file per branch in the list in order to use the stats file that was current as of a snapshot. But there is a benefit to this as well because you can always calculate old stats files and add them. For example, you can calculate stats for all of the tags in a table after they are tagged, even if the tagged snapshot itself didn't have stats originally.

Over all, I'm more and more confident that a top-level `stats` field is the right way to go.

Reply:
+1 for attaching stats file to table level, stats data needs to be re-calculated for each snapshot that insert or update data, the history stats files is only used for time travel query and not useful next calculation, it seems to be lower value.

Reply:
If this go into table metadata, how should i find out stat file appropriate to read from some previous table snapshot?
Should we keep multiple entries with `"ref": "main"` and basically pick the latest one with `value of "last-updated-snapshot-id" < selected_snapshot_id`?

Also, when updating the stats, how do I select which stat file to update from (which is current, and which historical)?
To avoid any confusion, we could make "ref", the branch name, as a key in the map, with value being an ordered list of stat file entries.

Reply:
I'm not really set on either location but I do have concerns like @findepi about issuing queries against older snapshots. Does this means we keep a history of stat files in the table metadata as well?

Reply:
I'm trying to think through the table level structure, I assume I need to be able to store data like
Stats files : Valid for Snapshots [x -> z]. It seems like we then have some difficulty in extending this range for commits which do not invalidate the snapshot unless the writer is clear on the stats file. A stats file tracking table wide aggregates could survive a `replace` but a bloom filter which pointed to individual data files would not.

I guess it is then still up to the writer of the new metadata.json to know whether or not the Stats file is valid for its new commit.

Reply:
@findepi, you would find the stats file written for the closest snapshot. That may be a snapshot that was written after: if a compaction runs and then stats are calculated for the compacted snapshot it would be closest for the snapshot that was compacted.

The file that should be updated is the newest ancestor with stats in a given snapshot's history. There are utils for traversing ancestors, so you'd do that until you find an ancestor with a stats file.

For the storage, I think we should use a list. Table metadata can be indexed however is convenient in memory. Adding requirements for ordering within the format is difficult (what happens when one is out of order?) and leads to issues with time skew.

@RussellSpitzer, I think it is simple to traverse ancestors to find the last stats file. And we can add stats files for older snapshots later if needed (the tagging case). I think it is actually much cleaner by not keeping these in snapshots.

Reply:
I don't think Snapshots is the right place to hold them @rdblue , I just am not sure how we are traversing ancestors in this case. Are you suggesting we have a list of stats files, each of which is associated with a single snapshot file. Then we we traverse our ancestors in the snapshot tree to find stats files? 

When we don't have a stats file for the current snapshot (or snapshot in question) do we return no stats files? I think that would make sense and instead we rely on a metadata only operation for associating old stats files with newer snapshots.

For example, older or non-puffin capable writer adds a snapshot B to existing table with Snapshot A

```
Table : { snaps {A, B} , stats { A -> A' } }
```

When reading B I think we have two options

1. Report there are no stats files associated with B
2. Use ancestor, find A -> A' and then attempt to determine if the operation to make B invalidated A'

I think 1. is probably safest and then we would add an operation for a client that understands A', `propagate_stats` or something which would add a new metadata.json
```
Table {snaps {A, B}, stats { A -> A', B -> A'}
```

Reply:
> We could design a new table-level structure that tracks stats files per branch, something like this:
> 
> ```json
> "stats": [
>     { "ref": "main", "last-updated-snapshot-id": 1, "stats-path": "...", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ] },
>     { "ref": "test", "last-updated-snapshot-id": 2, "stats-path": "...", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ] }
>   ]
> ```

@rdblue what do we need "ref" for?

Reply:
+1 for the idea to detach stats files from a snapshot. This is due to immutability of snapshot, which means any internal state change of a snapshot will have to create a new snapshot. We can update/generate stats files without creating a new snapshot by detaching them.

We can still consider stats file as an external property of snapshot, so that stats files are indexed by snapshot. So that we search the stats files by searching the snapshot history.
```
"stats": [
    { "snapshot-id": 1, stats-files[{"stats-path": "path/file1", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ]},{"stats-path": "path/file2", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ]}]},
    { "snapshot-id": 2, stats-files[{"stats-path": "path/file3", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ]},{"stats-path": "path/file4", "file-size-in-bytes": 1024, "footer-size-in-bytes": 100, "blob-metadata": [ {}, {} ]}]},
  ]
```

Reply:
@flyrain, I think there should still be just one stats file per snapshot. It is easy to merge them and we don't want to accumulate them.

Reply:
@findepi, do you think you can update this PR to add the `stats` structure as discussed here?

Reply:
Updated accordingly.